Fix BLA updates in Register and RegisterArray

SilvrrLining opened the pull request:
updated buffers.py in meeting.

SilvrrLining committed:
Update buffers.py

cmekik commented on pyClarion/components/buffers.py:
```diff
@@ -381,13 +385,21 @@ def call(self, inputs):
+                # Remove items below threshold.
+                drop = [x for x in cell.store if self.blas[x].below_threshold]
+                cell.store.drop(keys=drop)
+            for x in drop:
+                del self.blas[x]
```
This line is prone to missing deletions. Should probably look something like this:

```suggestion
            for x, bla in self.blas.items():
                if bla.below_threshold:
                    del self.blas[x]
```

cmekik commented on pyClarion/components/buffers.py:
```diff
@@ -227,21 +230,21 @@ def call(self, inputs):
             drop = [x for x in self.store if self.blas[x].below_threshold]
             self.store.drop(keys=drop)
             for x in drop:
-                del self.blas[item]
+                del self.blas[x]
```
See comment on lines +394 to +395. A similar change is needed here.

cmekik reviewed (dismissed):
Thanks for the PR!

We forgot to disable BLA updates for sub-registers in `RegisterArray.__init__()`. 

There are also a couple of other tweaks that I spotted.

SilvrrLining committed:
Apply suggestions from code review

Co-authored-by: Can Mekik <can.mekik@gmail.com>

cmekik commented:
I pushed some updates to dev which affect this a little bit:
- I rewrote a lot of other code in `buffers.py`, so the file is now conflicted
- There is now a new `BLAs.prune()` method, which may come in handy with the fix
- In `RegisterArray`, sub-registers are now defined in `interface()` instead of in `__init__()`

I imagine the solution will look something like this:
- Provide an `update_blas` option, which is `True` by default, in both `Register` and `RegisterArray`. 
- If the `update_blas` option is `False`, the BLA database should not be stepped or pruned in the `call()` method of either class. 
- In `Register.call()`, if `update_blas` is `False` but a BLA database is available, items below threshold should still be removed from the register store.  
- When initializing register cells in `RegisterArray.interface()`, pass in the BLA database but set `update_blas` to `False`. This way, when an object is written to the register cell, it will record the invocation.

SilvrrLining committed:
Merge remote-tracking branch 'upstream/dev' into dev

SilvrrLining committed:
fixed bugs in buffers, added a new method in blas

SilvrrLining committed:
fixed spelling in buffers

SilvrrLining committed:
Merge remote-tracking branch 'upstream/dev' into dev

SilvrrLining committed:
fixed blas and buffers bug

needs further thinking

cmekik commented:
Thanks for the updates! 

After studying this and thinking a bit more, I'm convinced that the BLA update sequence for registers has to be as below, all done within one simulation step.

1. Register any new invocations
2. Step the BLA database
3. Drop any items below threshold from store
4. Prune the BLA database

The current design seems flawed because there isn't a clean way to achieve this ordering of events for every subordinate register if BLA database stepping and pruning are controlled by `RegisterArray` but invocation registration and item clearing is controlled by subordinate registers. 
 
So, I propose the following:

- Replace the option `update_blas` with one called `shares_blas` (default is `False`). Here is a docstring for this parameter.
```python
"""
:param shares_blas: If True, will only report invocations. Otherwise, 
    will also maintain BLA database and remove items in store that are 
    below threshold.    
"""
```
- In `Register.call()`, insert the following block just prior to return. This replaces previous code associated with maintaining the BLA database and store (but not BLA invocation code).
```python
if self.blas is not None and not self.shares_blas:
    self.blas.step()
    self.store.drop(keys=self.blas.keys_below_threshold(self.store))
    self.blas.prune()
```
- In `RegisterArray.call()`, insert the following block just prior to return in `call()`. This replaces previous code associated with maintaining the BLA database.
```python
if self.blas is not None and not self.shares_blas:
    self.blas.step()
    for cell in self.cells:
        cell.store.drop(keys=self.blas.keys_below_threshold(cell.store))
    self.blas.prune()
```
- Finally, in `RegisterArray.interface()`, pass `shares_blas=True` to all subordinate registers.

SilvrrLining committed:
Fixed update_blas error in Buffer

- changed update_blas to shares_blas and default it as negative
- follow the steps
1. Register any new invocations
2. Step the BLA database
3. Drop any items below threshold from store
4. Prune the BLA database

cmekik commented:
Thank you for the fix! Looks good.
